pgxn/neon/walproposer.c

```diff
@@ -1335,76 +1293,101 @@ SendAppendRequests(Safekeeper *sk)
 			return sk->state == SS_ACTIVE;
 
 		/* Event set will be updated in the end of HandleActiveState */
-		sk->flushWrite = false;
+		sk->active_state = SS_ACTIVE_SEND;
 	}
 
 	while (sk->streamingAt != wp->availableLsn || !sentAnything)
 	{
-		sentAnything = true;
+		if (sk->active_state == SS_ACTIVE_SEND)
```

Comment:
It's not good that we don't send zero-length `AppendRequest`'s keepalives if we're now waiting for the WAL in `SS_ACTIVE_READ_WAL` state.

Reply:
Well, it is possible to implement that, though not trivial as sending keepalive might block. But do we need it? If we are in SS_ACTIVE_READ_WAL, sk is lagging, so it should receive next chunk very soon, or something is really broken and connection will be dead soon regardless of keepalives.

Reply:
AFAIU if donor is slow to send WAL, then sk connection will become stale without keepalives, and can be killed by the safekeeper.